src/manifest.cpp

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2020 YADRO
+
+#include "manifest.hpp"
+
+#include <limits.h>
+#include <unistd.h>
+
+#include <fstream>
+#include <regex>
+
+namespace fs = std::filesystem;
+
+/** @brief Name of the manifest file. */
+static const std::string manifestFile = "bmc.manifest";
+
+/** @brief Name of OS version property. */
+static const std::string osVersionProp = "VERSION";
+/** @brief Name of machine (platform) name property. */
+static const std::string machineNameProp = "MACHINE";
+/** @brief Name of host name property. */
+static const std::string hostNameProp = "HOSTNAME";
+
+/**
+ * @brief Parse ini file.
+ *
+ * @param[in] iniFile path to the ini file to parse
+ *
+ * @throw std::runtime_error in case of errors
+ *
+ * @return map of ini values
+ */
+static std::map<std::string, std::string> parseIni(const fs::path& iniFile)
+{
+    std::map<std::string, std::string> data;
+    const std::regex iniRegex("([^= ]+)\\s*=\\s*\"?([^\"]+)\"?");
+
+    std::ifstream file(iniFile);
+    if (!file)
+    {
+        std::string err = "Error opening file ";
+        err += iniFile;
+        throw std::runtime_error(err);
+    }
+    std::string line;
+    while (std::getline(file, line))
+    {
+        std::smatch match;
+        if (std::regex_match(line, match, iniRegex))
+        {
+            const auto& name = match[1].str();
+            const auto& value = match[2].str();
+            data.insert(std::make_pair(name, value));
+        }
+    }
+
+    return data;
+}
+
+Manifest::Manifest(const std::filesystem::path& rootFs)
+{
+    const fs::path osRelease = rootFs / "etc/os-release";
+    const std::map<std::string, std::string> ini = parseIni(osRelease);
+    const std::map<std::string, std::string> osr{
+        {"OPENBMC_TARGET_MACHINE", machineNameProp},
+        {"VERSION", osVersionProp}};
+    for (const auto& it : osr)
+    {
+        const auto val = ini.find(it.first);
+        if (val == ini.end())
+        {
+            std::string err = "Invalid os-release file format: Property ";
+            err += it.first;
+            err += " not found in file ";
+            err += osRelease;
+            throw std::runtime_error(err);
+        }
+        properties.insert(std::make_pair(it.second, val->second));
+    }
+
+    char hostname[HOST_NAME_MAX];
+    if (gethostname(hostname, sizeof(hostname)) != 0)
+    {
+        throw std::runtime_error("Unable to get host name");
+    }
+    properties.insert(std::make_pair(hostNameProp, hostname));
+}
+
+Manifest Manifest::load(const fs::path& dir)
+{
+    Manifest manifest;
+
+    const fs::path mnfFile = dir / manifestFile;
+    const std::map<std::string, std::string> ini = parseIni(mnfFile);
```

Comment:
This will create a hard link between the os-release file and manifest format. I don't like it. Also, I think we should not store the fields that are not handled by the program.

Reply:
I agree with Artem here